lib/trivial.nix

```
@@ -1004,11 +1004,18 @@ in
 
     : 2\. Function argument
   */
-  setFunctionArgs = f: args: {
-    # TODO: Should we add call-time "type" checking like built in?
-    __functor = self: f;
-    __functionArgs = args;
-  };
+  setFunctionArgs =
```

```suggestion
setFunctionArgs = f: args:
 if f ? __functor then
    f // { __functionArgs = args; }
 else if isAttrs f then
    f // {
      __functor = self: f;
      __functionArgs = args;
    }
 else 
    {
      __functor = self: f;
      __functionArgs = args;
    };
```

please also add some tests in misc.nix

---

The impact and the motivation behind this change is a bit unclear to me right now. Could you elaborate?